File: problem_solutions/fibonacci-number/iterative_solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int fib(int n) {
        // LeetCode constrains n to 0..30
        int fibVals[31];

        // equivalent to the base case in the recursive solution
        fibVals[0] = 0;
        fibVals[1] = 1;

        for (int i = 2; i <= n; ++i) {
            // transition
            fibVals[i] = fibVals[i - 1] + fibVals[i - 2];
        }

        return fibVals[n];
    }
};

int main() {
    Solution sol;
    int n = 7;
    cout << sol.fib(n) << endl;

    return 0;
}

File: problem_solutions/fibonacci-number/iterative_solution.py
class Solution:
    def fib(self, n: int) -> int:
        # LeetCode constrains n to 0..30
        fib_vals = [0] * 31

        # equivalent to the base case in the recursive solution
        fib_vals[0] = 0
        fib_vals[1] = 1

        for i in range(2, n + 1):
            # transition
            fib_vals[i] = fib_vals[i - 1] + fib_vals[i - 2]

        return fib_vals[n]


if __name__ == "__main__":
    sol = Solution()
    n = 7
    print(sol.fib(n))
